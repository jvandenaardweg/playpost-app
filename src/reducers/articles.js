export const GET_ARTICLES = 'articles/LOAD';
export const GET_ARTICLES_SUCCESS = 'articles/LOAD_SUCCESS';
export const GET_ARTICLES_FAIL = 'articles/LOAD_FAIL';

export function articlesReducer(state = { articles: [] }, action) {
  switch (action.type) {
    case GET_ARTICLES:
      return { ...state, isLoading: true };
    case GET_ARTICLES_SUCCESS:
      return { ...state, isLoading: false, articles: action.payload.data };
    case GET_ARTICLES_FAIL:
      return {
        ...state,
        isLoading: false,
        error: 'Error while fetching the articles'
      };
    default:
      return state;
  }
}

export function listArticles(user) {
  return {
    type: GET_ARTICLES,
    payload: {
      request: {
        url: `/users/${user}/repos`
      }
    }
  };
}
