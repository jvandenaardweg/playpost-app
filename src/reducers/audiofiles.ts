
export const SET_DOWNLOADED_AUDIOFILE = 'audiofiles/SET_DOWNLOADED_AUDIOFILE';
export const RESET_STATE = 'audiofiles/RESET_STATE';

export type AudiofilesState = Readonly<{
  isLoading: boolean;
  downloaded: ReadonlyArray<Api.Audiofile>;
  error: string;
}>;

export const initialState: AudiofilesState = {
  isLoading: false,
  downloaded: [],
  error: ''
};

/* tslint:disable-next-line no-any */
export function audiofilesReducer(state = initialState, action: any): AudiofilesState {
  switch (action.type) {
    case SET_DOWNLOADED_AUDIOFILE:
      const audiofile: Api.Audiofile = action.payload;

      return {
        ...state,
        isLoading: false,
        downloaded: [
          ...state.downloaded.slice(0, 0),
          audiofile,
          ...state.downloaded.slice(0)
        ],
        error: ''
      };

    case RESET_STATE:
      return {
        ...initialState
      };

    default:
      return state;
  }
}

export const resetAudiofilesState = () => ({
  type: RESET_STATE
});

export const setDownloadedAudiofile = (audiofile: Api.Audiofile) => ({
  type: SET_DOWNLOADED_AUDIOFILE,
  payload: audiofile
});
