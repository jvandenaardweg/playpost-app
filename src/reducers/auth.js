import Analytics from 'appcenter-analytics';

export const POST_AUTH = 'auth/POST_AUTH';
export const POST_AUTH_SUCCESS = 'auth/POST_AUTH_SUCCESS';
export const POST_AUTH_FAIL = 'auth/POST_AUTH_FAIL';
export const REMOVE_AUTH = 'auth/REMOVE_AUTH';
export const SET_AUTH_TOKEN = 'auth/SET_AUTH_TOKEN';

const POST_AUTH_FAIL_MESSAGE = 'An unknown error happened while loggin you in. Please contact us when this happens all the time.';

const initialState = {
  isLoading: false,
  token: null,
  error: null
};

export function authReducer(state = initialState, action) {
  switch (action.type) {
    case POST_AUTH:
      return {
        ...state,
        isLoading: true
      };

    case POST_AUTH_SUCCESS:
      Analytics.trackEvent('Auth success');

      return {
        ...state,
        isLoading: false,
        token: action.payload.data.token,
        error: null
      };

    case POST_AUTH_FAIL:
      if (action.error.response && action.error.response.data && action.error.response.data.message) {
        Analytics.trackEvent('Error auth', { message: action.error.response.data.message });
      } else {
        Analytics.trackEvent('Error auth', { message: POST_AUTH_FAIL_MESSAGE });
      }

      return {
        ...state,
        isLoading: false,
        token: null,
        error: (action.error.response) ? action.error.response.data.message : POST_AUTH_FAIL_MESSAGE
      };

    case REMOVE_AUTH:
      return {
        ...initialState
      };

    case SET_AUTH_TOKEN:
      return {
        ...state,
        token: action.payload.token
      };

    default:
      return state;
  }
}

export function setAuthToken(token) {
  return {
    type: SET_AUTH_TOKEN,
    payload: {
      token
    }
  };
}

export function removeAuth() {
  return {
    type: REMOVE_AUTH
  };
}

export function postAuth(email, password) {
  return {
    type: POST_AUTH,
    payload: {
      request: {
        method: 'post',
        url: '/v1/auth',
        data: {
          email,
          password
        }
      }
    }
  };
}
