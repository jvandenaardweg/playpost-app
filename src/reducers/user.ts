import DeviceInfo from 'react-native-device-info';

import {
  CREATE_USER_FAIL_MESSAGE,
  DELETE_USER_FAIL_MESSAGE,
  GENERIC_NETWORK_ERROR,
  GET_USER_FAIL_MESSAGE,
  PATCH_USER_FAIL_MESSAGE,
  SAVE_SELECTED_VOICE_FAIL_MESSAGE
} from '../constants/messages';

export const GET_USER = 'user/GET_USER';
export const GET_USER_SUCCESS = 'user/GET_USER_SUCCESS';
export const GET_USER_FAIL = 'user/GET_USER_FAIL';

export const PATCH_USER = 'user/PATCH_USER';
export const PATCH_USER_SUCCESS = 'user/PATCH_USER_SUCCESS';
export const PATCH_USER_FAIL = 'user/PATCH_USER_FAIL';

export const CREATE_USER = 'user/CREATE_USER';
export const CREATE_USER_SUCCESS = 'user/CREATE_USER_SUCCESS';
export const CREATE_USER_FAIL = 'user/CREATE_USER_FAIL';

export const DELETE_USER = 'user/DELETE_USER';
export const DELETE_USER_SUCCESS = 'user/DELETE_USER_SUCCESS';
export const DELETE_USER_FAIL = 'user/DELETE_USER_FAIL';

export const SAVE_SELECTED_VOICE = 'user/SAVE_SELECTED_VOICE';
export const SAVE_SELECTED_VOICE_SUCCESS = 'user/SAVE_SELECTED_VOICE_SUCCESS';
export const SAVE_SELECTED_VOICE_FAIL = 'user/SAVE_SELECTED_VOICE_FAIL';
export const RESET_SAVE_SELECTED_VOICE_ERROR = 'user/RESET_SAVE_SELECTED_VOICE_ERROR';


export const SET_PLAYBACK_SPEED = 'user/SET_PLAYBACK_SPEED';

export const RESET_USER_STATE = 'user/RESET_USER_STATE';
export const RESET_USER_ERROR = 'user/RESET_USER_ERROR';

export type UserState = Readonly<{
  isLoading: boolean;
  isLoadingDelete: boolean;
  isLoadingPatchUser: boolean;
  isLoadingSaveSelectedVoice: boolean;
  details: Api.User | null;
  error: string;
  errorSaveSelectedVoice: string;
  deviceLocale: string;
  playbackSpeed: number;
}>;

export const initialState: UserState = {
  isLoading: false,
  isLoadingDelete: false,
  isLoadingPatchUser: false,
  isLoadingSaveSelectedVoice: false,
  details: null,
  error: '',
  errorSaveSelectedVoice: '',
  deviceLocale: DeviceInfo.getDeviceLocale(),
  playbackSpeed: 1
};

/* tslint:disable no-any */
export function userReducer(state = initialState, action: any): UserState {
  switch (action.type) {
    case GET_USER:
      return {
        ...state,
        isLoading: true,
        error: ''
      };

    case GET_USER_SUCCESS:
      return {
        ...state,
        isLoading: false,
        details: action.payload.data,
        error: ''
      };

    case GET_USER_FAIL:
      let getUserFailMessage = '';

      // Network error
      if (action.error && action.error.status === 0) {
        getUserFailMessage = GENERIC_NETWORK_ERROR;
      } else {
        // Error, from the API
        if (action.error && action.error.response && action.error.response.data && action.error.response.data.message) {
          getUserFailMessage = action.error.response.data.message;
        } else {
          getUserFailMessage = GET_USER_FAIL_MESSAGE;
        }
      }

      return {
        ...state,
        isLoading: false,
        error: getUserFailMessage
      };

    case CREATE_USER:
      return {
        ...state,
        isLoading: true,
        error: ''
      };

    case CREATE_USER_SUCCESS:
      return {
        ...state,
        isLoading: false,
        details: action.payload.data,
        error: ''
      };

    case CREATE_USER_FAIL:
      let createUserFailMessage = '';

      // Network error
      if (action.error && action.error.status === 0) {
        createUserFailMessage = GENERIC_NETWORK_ERROR;
      } else {
        // Error, from the API
        if (action.error && action.error.response && action.error.response.data && action.error.response.data.message) {
          createUserFailMessage = action.error.response.data.message;
        } else {
          createUserFailMessage = CREATE_USER_FAIL_MESSAGE;
        }
      }

      return {
        ...state,
        isLoading: false,
        details: null,
        error: createUserFailMessage
      };

    case DELETE_USER:
      return {
        ...state,
        isLoadingDelete: true,
        error: ''
      };

    case DELETE_USER_SUCCESS:
      return {
        ...state,
        isLoadingDelete: false,
        error: ''
      };

    case DELETE_USER_FAIL:
      let deleteUserFailMessage = '';

      // Network error
      if (action.error && action.error.status === 0) {
        deleteUserFailMessage = GENERIC_NETWORK_ERROR;
      } else {
        // Error, from the API
        if (action.error && action.error.response && action.error.response.data && action.error.response.data.message) {
          deleteUserFailMessage = action.error.response.data.message;
        } else {
          deleteUserFailMessage = DELETE_USER_FAIL_MESSAGE;
        }
      }

      return {
        ...state,
        isLoadingDelete: false,
        error: deleteUserFailMessage
      };

    case PATCH_USER:
      return {
        ...state,
        isLoadingPatchUser: true,
        error: ''
      };

    case PATCH_USER_SUCCESS:
      return {
        ...state,
        isLoadingPatchUser: false,
        error: ''
      };

    case PATCH_USER_FAIL:
      let updateUserEmailFailMessage = '';

      // Network error
      if (action.error && action.error.status === 0) {
        updateUserEmailFailMessage = GENERIC_NETWORK_ERROR;
      } else {
        // Error, from the API
        if (action.error && action.error.response && action.error.response.data && action.error.response.data.message) {
          updateUserEmailFailMessage = action.error.response.data.message;
        } else {
          updateUserEmailFailMessage = PATCH_USER_FAIL_MESSAGE;
        }
      }

      return {
        ...state,
        isLoadingPatchUser: false,
        error: updateUserEmailFailMessage
      };

    case SAVE_SELECTED_VOICE:
      return {
        ...state,
        isLoadingSaveSelectedVoice: true,
        errorSaveSelectedVoice: ''
      };

    case SAVE_SELECTED_VOICE_SUCCESS:
      return {
        ...state,
        isLoadingSaveSelectedVoice: false,
        errorSaveSelectedVoice: ''
      };

    case SAVE_SELECTED_VOICE_FAIL:
      let saveSelectedVoiceFailMessage = '';

      // Network error
      if (action.error && action.error.status === 0) {
        saveSelectedVoiceFailMessage = GENERIC_NETWORK_ERROR;
      } else {
        // Error, from the API
        if (action.error && action.error.response && action.error.response.data && action.error.response.data.message) {
          saveSelectedVoiceFailMessage = action.error.response.data.message;
        } else {
          saveSelectedVoiceFailMessage = SAVE_SELECTED_VOICE_FAIL_MESSAGE;
        }
      }

      return {
        ...state,
        isLoadingSaveSelectedVoice: false,
        errorSaveSelectedVoice: saveSelectedVoiceFailMessage
      };

    case SET_PLAYBACK_SPEED:
      return {
        ...state,
        playbackSpeed: action.payload
      };

    case RESET_SAVE_SELECTED_VOICE_ERROR:
      return {
        ...state,
        errorSaveSelectedVoice: ''
      };

    case RESET_USER_ERROR:
      return {
        ...state,
        error: ''
      };

    case RESET_USER_STATE:
      return {
        ...initialState
      };

    default:
      return state;
  }
}

export function resetUserState() {
  return {
    type: RESET_USER_STATE
  };
}

export function resetSaveSelectedVoiceError() {
  return {
    type: RESET_SAVE_SELECTED_VOICE_ERROR
  };
}

export function resetUserError() {
  return {
    type: RESET_USER_ERROR
  };
}

export function getUser() {
  return {
    type: GET_USER,
    payload: {
      request: {
        method: 'get',
        url: '/v1/me'
      }
    }
  };
}

export function patchUser({ email, password }: { email?: string, password?: string }) {
  return {
    type: PATCH_USER,
    payload: {
      request: {
        method: 'patch',
        url: '/v1/me',
        data: {
          email,
          password
        }
      }
    }
  };
}

export function createUser(email: string, password: string) {
  return {
    type: CREATE_USER,
    payload: {
      request: {
        method: 'post',
        url: '/v1/users',
        data: {
          email,
          password
        }
      }
    }
  };
}

export function deleteUser() {
  return {
    type: DELETE_USER,
    payload: {
      request: {
        method: 'delete',
        url: '/v1/me'
      }
    }
  };
}

/**
 * Saves the selected voice as a default voice for the voice's language
 *
 * @param voiceId
 */
export function saveSelectedVoice(voiceId: string) {
  return {
    type: SAVE_SELECTED_VOICE,
    payload: {
      request: {
        method: 'post',
        url: '/v1/me/voices',
        data: {
          voiceId
        }
      }
    }
  };
}

export function setPlaybackSpeed(speed: number) {
  return {
    type: SET_PLAYBACK_SPEED,
    payload: speed
  };
}
