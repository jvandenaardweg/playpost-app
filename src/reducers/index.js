import { articlesReducer } from './articles';
import { playerReducer } from './player';
import { authReducer } from './auth';
import { usersReducer } from './users';

const rootReducer = {
  articles: articlesReducer,
  player: playerReducer,
  auth: authReducer,
  users: usersReducer
};

export default rootReducer;
