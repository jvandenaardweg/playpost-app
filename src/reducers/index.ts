import { combineReducers, Action } from 'redux';
import { AxiosResponse, AxiosRequestConfig } from 'axios';

import { playerReducer } from './player';
import { authReducer } from './auth';
import { userReducer } from './user';
import { playlistReducer } from './playlist';
import { audiofilesReducer } from './audiofiles';
import { voicesReducer } from './voices';

/* tslint:disable no-any */
export interface AxiosActionResponse extends AxiosResponse {
  payload: {
    data: any
  };
}

export interface AxiosAction extends Action {
  payload: {
    request: AxiosRequestConfig
  };
}

const rootReducer = combineReducers({
  player: playerReducer,
  auth: authReducer,
  user: userReducer,
  playlist: playlistReducer,
  audiofiles: audiofilesReducer,
  voices: voicesReducer
});

export type RootState = ReturnType<typeof rootReducer>;

export { rootReducer };
