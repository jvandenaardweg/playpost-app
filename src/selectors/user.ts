import { Platform } from 'react-native';
import { createSelector } from 'reselect';
import { SUBSCRIPTION_NAME, SUBSCRIPTION_PRODUCT_ID_FREE } from '../constants/in-app-purchase';
import { RootState } from '../reducers';
import { UserState } from '../reducers/user';
import { createDeepEqualSelector } from './index';

export interface UserSelectedVoiceByLanguageName {
  [key: string]: Api.Voice
}

export const userSelector = (state: RootState): UserState => state.user;

export const selectUserError = createSelector(
  [userSelector],
  user => user.error
);

export const selectUserErrorSaveSelectedVoice = createSelector(
  [userSelector],
  user => user.errorSaveSelectedVoice
);

export const selectUserIsLoading = createSelector(
  [userSelector],
  user => user.isLoading
);

export const selectUserDetails = createSelector(
  [userSelector],
  user => user.details
);

export const selectUserUsedInAppSubscriptionTrials = createSelector(
  [selectUserDetails],
  userDetails => userDetails && userDetails.usedInAppSubscriptionTrials
);

export const selectDeviceLocale = createSelector(
  [userSelector],
  user => {
    if (!user.deviceLocale) {
      return '';
    }

    // device locale could be: "en-NL" or "en"
    // So we always split the dash
    // Even if the dash does not exist, we return the first in the array
    const deviceLocaleSplitted = user.deviceLocale.split('-');

    if (!deviceLocaleSplitted || !deviceLocaleSplitted[0]) {
      return '';
    }

    return deviceLocaleSplitted[0];
  }
);

export const selectUserSelectedVoices = createDeepEqualSelector(
  [selectUserDetails],
  (userDetails) => {
    if (!userDetails || !userDetails.voiceSettings || !userDetails.voiceSettings.length) {
      return [];
    }
    return userDetails.voiceSettings.map(userVoiceSetting => userVoiceSetting.voice);
  }
);

export const selectUserSelectedVoiceByLanguageName = createDeepEqualSelector(
  [selectUserSelectedVoices],
  (voices): UserSelectedVoiceByLanguageName | null => {
    // Convert the array to an object
    // So we can easily pick a language inside our components
    const voicesObjectWithLanguageNameKeys: UserSelectedVoiceByLanguageName = voices.reduce((prev, curr) => {
      prev[curr.language.name] = curr;

      return prev;
    }, {})

    if (!Object.keys(voicesObjectWithLanguageNameKeys).length) {
      return null;
    }

    return voicesObjectWithLanguageNameKeys;
  }
);

export const selectUserSubscriptions = createDeepEqualSelector(
  [selectUserDetails],
  userDetails => userDetails && userDetails.inAppSubscriptions
);

export const selectUserIsSubscribed = createDeepEqualSelector(
  [selectUserDetails],
  (userDetails): boolean => !!(userDetails && userDetails.activeUserInAppSubscription)
);

/**
 * The active subscription's productId is different per platform.
 *
 */
export const selectUserActiveSubscriptionProductId = createDeepEqualSelector(
  [selectUserDetails, selectUserIsSubscribed],
  (userDetails, isSubscribed): string => {
    if (!isSubscribed) { return SUBSCRIPTION_PRODUCT_ID_FREE; }
    if (!userDetails) { return SUBSCRIPTION_PRODUCT_ID_FREE; }
    if (!userDetails.activeUserInAppSubscription) { return SUBSCRIPTION_PRODUCT_ID_FREE; }
    if (!userDetails.activeUserInAppSubscription.inAppSubscription) { return SUBSCRIPTION_PRODUCT_ID_FREE; }
    return userDetails.activeUserInAppSubscription.inAppSubscription.productId
  }
);

export const selectActiveUserInAppSubscription = createDeepEqualSelector(
  [selectUserDetails, selectUserIsSubscribed],
  (userDetails, isSubscribed): Api.UserInAppSubscriptionApple | Api.UserInAppSubscriptionGoogle | null => {
    if (!isSubscribed) { return null; }
    if (!userDetails) { return null; }
    if (!userDetails.activeUserInAppSubscription) { return null; }

    return userDetails.activeUserInAppSubscription;
  }
);

export const selectUserActiveSubscriptionName = createDeepEqualSelector(
  [selectActiveUserInAppSubscription],
  (userActiveSubscription): string => {
    if (!userActiveSubscription) {
      return SUBSCRIPTION_NAME[SUBSCRIPTION_PRODUCT_ID_FREE];
    }
    return userActiveSubscription.inAppSubscription.name
  }
);

export const selectUserIsEligibleForTrial = createDeepEqualSelector(
  [selectUserUsedInAppSubscriptionTrials],
  (usedInAppSubscriptionTrials): boolean => {
    if (!usedInAppSubscriptionTrials || !usedInAppSubscriptionTrials.length) {
      return true;
    }

    // If the platform is Android, look up if the user had a Google subscription trial before
    if (Platform.OS === 'android') {
      return !!!usedInAppSubscriptionTrials.find(usedInAppSubscription => usedInAppSubscription.service === 'google')
    }

    // If the platform is else (iOS), look up if the user had a Apple subscription trial before
    return !!!usedInAppSubscriptionTrials.find(usedInAppSubscription => usedInAppSubscription.service === 'apple')
  }
);

export const selectUserHasSubscribedBefore = createDeepEqualSelector(
  [selectUserSubscriptions],
  userInAppSubscriptions => {
    if (!userInAppSubscriptions || !userInAppSubscriptions.length) {
      return false;
    }

    // If the user has a previous in app subscription record, return true
    return !!userInAppSubscriptions.length;
  }
);

export const selectUserPlaybackSpeed = createSelector(
  [userSelector],
  user =>  user.playbackSpeed
)
