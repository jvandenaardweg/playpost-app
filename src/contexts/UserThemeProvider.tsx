import React, { useEffect, useState, ReactNode } from 'react';
import { Platform, StatusBar } from 'react-native';
import { eventEmitter, useDarkMode } from 'react-native-dark-mode'
// tslint:disable-next-line: no-submodule-imports
import { Mode } from 'react-native-dark-mode/dist/types';
import { connect } from 'react-redux';

import colors from '../constants/colors';
import { RootState } from '../reducers';
import { setUserSelectedTheme, userDefaultTheme, UserTheme } from '../reducers/user';
import { selectUserSelectedTheme } from '../selectors/user';

export const UserThemeContext = React.createContext<{ theme: UserTheme; }>({
  theme: userDefaultTheme // The default theme for new users is Light
});

export const UserThemeConsumer = UserThemeContext.Consumer;

interface IProps {
  children?: ReactNode;
}

export type Props = IProps & StateProps & DispatchProps;

export const UserThemeProviderComponent: React.FC<Props> = React.memo((props) => {
  const [localTheme, setLocalTheme] = useState(userDefaultTheme) // Default theme of the app is light
  const isDarkMode = useDarkMode();

  const handleModeChange = (newMode: Mode) => {
    // If the user's theme settings is set to auto, listen for mode changes and apply those changes
    // These changes happen when the user's OS switches from light to dark, and the other way around
    if (props.userSelectedTheme === UserTheme.auto) {
      switchTheme(newMode === 'dark' ? UserTheme.dark : UserTheme.light)
    }
  }

  const switchTheme = (theme: UserTheme) => {
    setStatusBarStyle(theme)

    // Use a local state so we can handle auto-theming based on the phone's OS theme
    return setLocalTheme(theme)
  }

  const setStatusBarStyle = (theme: UserTheme) => {
    if (theme === UserTheme.dark) {
      StatusBar.setBarStyle('light-content')

      if (Platform.OS === 'android') {
        StatusBar.setBackgroundColor(colors.pureBlack)
      }
    } else {
      StatusBar.setBarStyle('dark-content')

      if (Platform.OS === 'android') {
        StatusBar.setBackgroundColor(colors.white)
      }
    }
  }

  // mount
  useEffect(() => {
    eventEmitter.on('currentModeChanged', handleModeChange)

    // unmount
    return () => {
      eventEmitter.off('currentModeChanged', handleModeChange)
    }
  }, [])

  useEffect(() => {
    // Upon app start, determine which theme to use
    if (props.userSelectedTheme === UserTheme.auto) {
      switchTheme(isDarkMode ? UserTheme.dark : UserTheme.light)
    } else {
      switchTheme(props.userSelectedTheme)
    }
  }, [props.userSelectedTheme, isDarkMode])

  return <UserThemeContext.Provider value={{ theme: localTheme }}>{props.children}</UserThemeContext.Provider>;
})

interface StateProps {
  userSelectedTheme: ReturnType<typeof selectUserSelectedTheme>;
}

interface DispatchProps {
  setUserSelectedTheme: typeof setUserSelectedTheme;
}

const mapStateToProps = (state: RootState): StateProps => ({
  userSelectedTheme: selectUserSelectedTheme(state)
});

const mapDispatchToProps = {
  setUserSelectedTheme
};

export const UserThemeProvider = connect(
  mapStateToProps,
  mapDispatchToProps
)(UserThemeProviderComponent);
