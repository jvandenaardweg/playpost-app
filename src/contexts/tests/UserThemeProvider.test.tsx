import React from 'react';
import { StatusBar } from 'react-native';
import { render, RenderAPI } from 'react-native-testing-library';
import { UserTheme } from '../../reducers/user';
import { Props, UserThemeConsumer, UserThemeProviderComponent } from '../UserThemeProvider';

const setUserSelectedThemeHandler = jest.fn();

const defaultProps: Props = {
  userSelectedTheme: UserTheme.light,
  setUserSelectedTheme: setUserSelectedThemeHandler
}
describe('UserThemeProvider', () => {

  describe('rendering', () => {
    let wrapper: RenderAPI;

    beforeEach(() => {
      jest.resetAllMocks();
      jest.restoreAllMocks();
    });

    beforeAll(() => {
      const props = {
        ...defaultProps
      }

      wrapper = render(<UserThemeProviderComponent {...props} />);
    });

    it('should render correctly', () => {
      expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should provide the correct default theme', () => {
      const props = {
        ...defaultProps
      }

      const spySetBarStyle = jest.spyOn(StatusBar, 'setBarStyle');

      wrapper = render(
        <UserThemeProviderComponent {...props}>
          <UserThemeConsumer>{context => context.theme}</UserThemeConsumer>
        </UserThemeProviderComponent>
      );

      expect(wrapper.toJSON()).toBe('light');

      expect(spySetBarStyle).toHaveBeenCalledTimes(1);
      expect(spySetBarStyle).toHaveBeenCalledWith('dark-content');
    });

    it('should provide the correct theme when changed', () => {
      const props = {
        ...defaultProps,
        userSelectedTheme: UserTheme.dark
      }

      const spySetBarStyle = jest.spyOn(StatusBar, 'setBarStyle');

      wrapper = render(
        <UserThemeProviderComponent {...props}>
          <UserThemeConsumer>{context => context.theme}</UserThemeConsumer>
        </UserThemeProviderComponent>
      );

      expect(wrapper.toJSON()).toBe('dark');
      expect(spySetBarStyle).toHaveBeenCalledTimes(1);
      expect(spySetBarStyle).toHaveBeenCalledWith('light-content');
    });

    it('should provide the correct theme when the theme is set on auto', () => {
      const props = {
        ...defaultProps,
        userSelectedTheme: UserTheme.auto
      }

      const spySetBarStyle = jest.spyOn(StatusBar, 'setBarStyle');

      wrapper = render(
        <UserThemeProviderComponent {...props}>
          <UserThemeConsumer>{context => context.theme}</UserThemeConsumer>
        </UserThemeProviderComponent>
      );

      // "light" is the default theme in the testing environment
      expect(wrapper.toJSON()).toBe('light');

      expect(spySetBarStyle).toHaveBeenCalledTimes(1);
      expect(spySetBarStyle).toHaveBeenCalledWith('dark-content');
    });
  });
});
