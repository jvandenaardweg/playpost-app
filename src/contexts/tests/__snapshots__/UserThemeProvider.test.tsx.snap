// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserThemeProvider rendering should render correctly 1`] = `null`;
